Staking contract

mtabasco opened the pull request:
Reference: https://hackmd.io/@mtabasco/ssv_staking

mtabasco committed:
feat:staking contract + cSSV WIP

mtabasco committed:
Merge branch 'feat/eth-eb-merge' into feat/staking-contract

mtabasco committed:
feat:update deploy script, remove withdrawNetworkEarnings, optimizations

venimir-ssv commented on contracts/modules/SSVStaking.sol:
```diff
@@ -0,0 +1,222 @@
+    }
+
+    function requestUnstake(uint256 amount) external {
+        if (amount == 0) revert ZeroAmount();
```
Shall we check if there is pending unstake process already active?

mtabasco replied:
yes, it's done here:
`if (s.withdrawals[msg.sender].amount != 0) revert CooldownActive();`

yurii-ssv committed:
fix: module names

venimir-ssv commented on contracts/libraries/SSVStorageStaking.sol:
```diff
@@ -0,0 +1,37 @@
+    /// @notice Amount of cSSV burned and pending to be withdrawn as SSV
+    uint192 amount;
+    /// @notice Timestamp after which the pending unstake can be withdrawn
+    uint64 unlockTime;
```
Shall we use block number instead of timestamp?

mtabasco replied:
let's keep timestamp as we use it on declaring new operator fees, etc

venimir-ssv commented on contracts/token/CSSVToken.sol:
```diff
@@ -0,0 +1,40 @@
+}
+
+contract CSSVToken is ERC20 {
+    error NotSSVStaking();
```
I think it's better to move the errors in ISSVStaking

mtabasco replied:
As this will be an isolated contract, I think it is good to not have interface dependencies with SSV Network contracts.

yurii-ssv commented on contracts/modules/SSVViews.sol:
```diff
@@ -463,6 +467,63 @@ contract SSVViews is ISSVViews {
+
+    function totalStaked() external view override returns (uint256) {
+        address cssv = SSVStorageStaking.load().cssv;
+        return cssv == address(0) ? 0 : ICSSVToken(cssv).totalSupply();
```
maybe it will be better to enforce saving cssv address as const/immutable in constructor/initializer to always be sure that it is not address(0)

mtabasco replied:
Yes, cSSV should be set when we upgrade the contract, so this check for address(0) can be removed. let me do it

mtabasco committed:
feat:updateable cooldownPeriod, remove address(0) check on cSSV

venimir-ssv committed:
docs: update ETH migration changelog from base to current HEAD (#341)

- Updated changelog to reflect all changes from commit a2e968f to bff3aed
- Added comprehensive documentation for SSV Staking Contract feature
- Added detailed file changes for staking-related contracts (SSVStaking, CSSVToken)
- Updated commit history with Phase 6: Staking Contract commits
- Updated statistics: 65 files changed, +15,708/-13,239 lines
- Removed obsolete references to withdrawNetworkEarnings (ETH version)
- Added staking migration path, security considerations, and testing recommendations
- Documented ETH earnings distribution through staking contract

yurii-ssv committed:
feat: add support for custom initializers

yurii-ssv committed:
chore: deployed addresses and abis

mtabasco committed:
sync base branch
